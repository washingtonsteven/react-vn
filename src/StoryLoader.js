import React, { Component } from "react";
import JSONFileInput from "./ui/JSONFileInput";
import "./StoryLoader.scss";
import ErrorMessage from "@@/ui/ErrorMessage";

class StoryLoader extends Component {
  state = { jsonURL: "data/story.json" };

  fileSelected = json => {
    if (!json.nodes) {
      this.setState(state => ({
        ...state,
        error: "Supplied JSON must have a `nodes` property"
      }));
    } else if (!json.meta) {
      this.setState(state => ({
        ...state,
        error: "Supplies JSON must have a `meta` property"
      }));
    } else {
      this.props.onFileLoaded && this.props.onFileLoaded(json);
    }
  };

  loadFile = e => {
    e.preventDefault();
    e.stopPropagation();
    this.props.onFilePathSet && this.props.onFilePathSet(this.state.jsonURL);
  };

  onLoadError = error => {
    this.setState(state => ({ ...state, error: error.message }));
  };

  setFilePath = e => {
    const jsonURL = e.target.value;
    this.setState(state => ({ ...state, jsonURL }));
  };

  clearError = () => this.setState(state => ({ ...state, error: null }));

  render() {
    return (
      <div className="story-loader">
        {!process.env.REACT_APP_IS_PLAYER && (
          <React.Fragment>
            <button onClick={this.props.onNew}>New Story</button>
            <div className="separator">Or</div>
          </React.Fragment>
        )}
        <div>
          <label htmlFor="story-file-loader">
            <span>Upload a .json story file</span>
            <JSONFileInput
              onFileSelected={this.fileSelected}
              onError={this.onLoadError}
              name="story-file-loader"
            />
          </label>
          {this.state.error && (
            <ErrorMessage onClose={this.clearError}>
              {this.state.error}
            </ErrorMessage>
          )}
        </div>
        <div className="separator">Or</div>
        <div>
          <form onSubmit={this.loadFile}>
            <label htmlFor="url-loader">
              <span>Enter a URL for a .json story file</span>
              <input
                type="text"
                placeholder="JSON url or serverpath"
                value={this.state.jsonURL}
                onChange={this.setFilePath}
                id="url-loader"
                name="url-loader"
              />
            </label>
            <button>Load!</button>
          </form>
        </div>
      </div>
    );
  }
}

export default StoryLoader;
