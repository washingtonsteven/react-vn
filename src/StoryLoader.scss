@import "~@@/_var";

.story-loader {
  label {
    width: 100%;
    padding: 5px 0;
    text-align: center;

    & > input {
      width: 60%;
      min-width: 500px;
      margin: 0 auto;
    }

    span.error {
      background-color: $red;
      border-color: $red;
      border-radius: 0;
      border-top-right-radius: 0.3rem;
      border-bottom-right-radius: 0.3rem;
      min-width: 30%;
    }
  }

  form {
    display: flex;
    grid-template-areas: "loader button";

    label {
      input {
        width: auto;
      }
    }

    button {
      grid-area: button;
      max-height: 36px;
      align-self: center;
      justify-self: left;
      margin: 0 0 0 10px;
    }
  }

  button {
    margin: 0 auto;
    display: block;
    color: $sky;
    border-color: $sky;

    &:hover {
      background-color: $sky;
      color: white;
    }
  }

  .separator {
    text-transform: uppercase;
    text-align: center;
    margin: 35px 0;
    font-size: 1.2rem;
    position: relative;

    &:before,
    &:after {
      content: "";
      display: block;
      position: absolute;
      top: 50%;
      transform: translate(0, -50%);
      width: 10%;
      height: 2px;
      background-color: #333;
    }

    &:before {
      left: 35%;
    }

    &:after {
      right: 35%;
    }
  }
}
