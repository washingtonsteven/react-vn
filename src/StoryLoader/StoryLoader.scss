@import "~@@/style/_var";

.story-loader {
  label {
    width: 100%;
    padding: 5px 0;
    text-align: center;
  }

  form {
    display: flex;
    grid-template-areas: "loader button";

    @media (max-width: $mobile_width) {
      display: block;
    }

    label {
      input {
        width: auto;
      }
    }

    button {
      grid-area: button;
      max-height: 36px;
      align-self: center;
      justify-self: left;
      margin: 0 0 0 10px;

      @media (max-width: $mobile_width) {
        margin-left: 0;
      }
    }
  }

  button {
    margin: 0 auto;
    display: block;
    color: $sky;
    border-color: $sky;

    &:hover {
      background-color: $sky;
      color: white;
    }
  }

  .separator {
    text-transform: uppercase;
    text-align: center;
    margin: 35px 0;
    font-size: 1.2rem;
    position: relative;

    &:before,
    &:after {
      content: "";
      display: block;
      position: absolute;
      top: 50%;
      transform: translate(0, -50%);
      width: 10%;
      height: 2px;
      background-color: #333;

      @media (max-width: $mobile_small) {
        display: none;
      }
    }

    &:before {
      left: 35%;
    }

    &:after {
      right: 35%;
    }
  }
}
