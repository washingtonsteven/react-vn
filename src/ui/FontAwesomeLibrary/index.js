import fontawesome from "@fortawesome/fontawesome";
import { faGithub, faTwitter } from "@fortawesome/fontawesome-free-brands";
import {
  faTrash,
  faExclamation,
  faTimesCircle
} from "@fortawesome/fontawesome-free-solid";

fontawesome.library.add(
  faTrash,
  faExclamation,
  faTimesCircle,
  faGithub,
  faTwitter
);
