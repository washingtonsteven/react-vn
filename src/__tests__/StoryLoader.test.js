import React from "react";
import FontAwesomeLibrary from "@@/ui/FontAwesomeLibrary";
import StoryLoader from "@@/StoryLoader";
import renderer from "react-test-renderer";

const e = {
  target: { value: "" },
  stopPropagation: jest.fn(),
  preventDefault: jest.fn()
};

let component = null;
let instance = null;
let tree = null;
const onFileLoaded = jest.fn();
const onFilePathSet = jest.fn();
const onNew = jest.fn();
const fnProps = { onFileLoaded, onFilePathSet, onNew };

beforeAll(() => {
  component = renderer.create(<StoryLoader {...fnProps} />);
  instance = component.getInstance();
  tree = component.toJSON();
});

it("renders without crashing", () => {
  expect(tree).toMatchSnapshot();
});

it("updates file url path", () => {
  const testURL = "test/test.json";
  instance.setFilePath({ target: { value: "test/test.json" } });
  expect(instance.state.jsonURL).toEqual(testURL);
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  const evt = { ...e };
  instance.loadFile(evt);
  expect(onFilePathSet).toBeCalledWith(testURL);
});

it("handles an load error from JSONFileInput", () => {
  const errorObj = { message: "Error form JSON File Input" };
  instance.onLoadError(errorObj);
  expect(instance.state.error).toEqual(errorObj.message);
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

describe("checks loaded json", () => {
  const noNodes = {};
  const noMeta = { nodes: [] };
  const valid = { nodes: [], meta: {} };

  it("Error if there are no nodes", () => {
    instance.fileSelected(noNodes);
    expect(instance.state.error).toBeTruthy();
    expect(instance.state.error).toMatch(/`nodes`/);
  });

  it("properly clears errors", () => {
    instance.clearError();
    expect(instance.state.error).toBeFalsy();
  });

  it("Error if there is no meta", () => {
    instance.fileSelected(noMeta);
    expect(instance.state.error).toBeTruthy();
    expect(instance.state.error).toMatch(/`meta`/);
  });

  it("Accepts JSON with nodes and meta set", () => {
    instance.fileSelected(valid);
    expect(onFileLoaded).toBeCalledWith(valid);
    expect(instance.state.error).toBeFalsy();
  });
});
