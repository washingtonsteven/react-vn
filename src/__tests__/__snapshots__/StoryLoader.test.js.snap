// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handles an load error from JSONFileInput 1`] = `
<div
  className="story-loader"
>
  <button
    onClick={[Function]}
  >
    New Story
  </button>
  <div
    className="separator"
  >
    Or
  </div>
  <div>
    <label
      htmlFor="story-file-loader"
    >
      <span>
        Upload a .json story file
      </span>
      <input
        accept="application/json"
        id="story-file-loader"
        name="story-file-loader"
        onChange={[Function]}
        type="file"
      />
    </label>
    <span
      className="error-message"
    >
      Error form JSON File Input
       
      <span
        className="close"
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-times-circle fa-w-16 "
          data-icon="times-circle"
          data-prefix="fas"
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm121.6 313.1c4.7 4.7 4.7 12.3 0 17L338 377.6c-4.7 4.7-12.3 4.7-17 0L256 312l-65.1 65.6c-4.7 4.7-12.3 4.7-17 0L134.4 338c-4.7-4.7-4.7-12.3 0-17l65.6-65-65.6-65.1c-4.7-4.7-4.7-12.3 0-17l39.6-39.6c4.7-4.7 12.3-4.7 17 0l65 65.7 65.1-65.6c4.7-4.7 12.3-4.7 17 0l39.6 39.6c4.7 4.7 4.7 12.3 0 17L312 256l65.6 65.1z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </span>
    </span>
  </div>
  <div
    className="separator"
  >
    Or
  </div>
  <div>
    <form
      onSubmit={[Function]}
    >
      <label
        htmlFor="url-loader"
      >
        <span>
          Enter a URL for a .json story file
        </span>
        <input
          id="url-loader"
          name="url-loader"
          onChange={[Function]}
          placeholder="JSON url or serverpath"
          type="text"
          value="test/test.json"
        />
      </label>
      <button>
        Load!
      </button>
    </form>
  </div>
</div>
`;

exports[`renders without crashing 1`] = `
<div
  className="story-loader"
>
  <button
    onClick={[Function]}
  >
    New Story
  </button>
  <div
    className="separator"
  >
    Or
  </div>
  <div>
    <label
      htmlFor="story-file-loader"
    >
      <span>
        Upload a .json story file
      </span>
      <input
        accept="application/json"
        id="story-file-loader"
        name="story-file-loader"
        onChange={[Function]}
        type="file"
      />
    </label>
  </div>
  <div
    className="separator"
  >
    Or
  </div>
  <div>
    <form
      onSubmit={[Function]}
    >
      <label
        htmlFor="url-loader"
      >
        <span>
          Enter a URL for a .json story file
        </span>
        <input
          id="url-loader"
          name="url-loader"
          onChange={[Function]}
          placeholder="JSON url or serverpath"
          type="text"
          value="data/story.json"
        />
      </label>
      <button>
        Load!
      </button>
    </form>
  </div>
</div>
`;

exports[`updates file url path 1`] = `
<div
  className="story-loader"
>
  <button
    onClick={[Function]}
  >
    New Story
  </button>
  <div
    className="separator"
  >
    Or
  </div>
  <div>
    <label
      htmlFor="story-file-loader"
    >
      <span>
        Upload a .json story file
      </span>
      <input
        accept="application/json"
        id="story-file-loader"
        name="story-file-loader"
        onChange={[Function]}
        type="file"
      />
    </label>
  </div>
  <div
    className="separator"
  >
    Or
  </div>
  <div>
    <form
      onSubmit={[Function]}
    >
      <label
        htmlFor="url-loader"
      >
        <span>
          Enter a URL for a .json story file
        </span>
        <input
          id="url-loader"
          name="url-loader"
          onChange={[Function]}
          placeholder="JSON url or serverpath"
          type="text"
          value="test/test.json"
        />
      </label>
      <button>
        Load!
      </button>
    </form>
  </div>
</div>
`;
